crates/dekaf/src/topology.rs

```diff
@@ -1,30 +1,104 @@
-use crate::connector::DeletionMode;
-use anyhow::Context;
-use futures::{StreamExt, TryStreamExt};
+use crate::{
+    connector::{DekafConfig, DekafResourceConfig, DeletionMode},
+    dekaf_shard_template_id, App, SessionAuthentication, TaskAuth, UserAuth,
+};
+use anyhow::{anyhow, bail, Context};
+use flow_client::fetch_task_authorization;
+use futures::{StreamExt, TryFutureExt, TryStreamExt};
 use gazette::{broker, journal, uuid};
+use itertools::Itertools;
+use models::RawValue;
 use proto_flow::flow;
+use std::time::Duration;
 
-/// Fetch the names of all collections which the current user may read.
-/// Each is mapped into a kafka topic.
-pub async fn fetch_all_collection_names(
-    client: &postgrest::Postgrest,
-) -> anyhow::Result<Vec<String>> {
-    #[derive(serde::Deserialize)]
-    struct Row {
-        catalog_name: String,
+impl UserAuth {
+    /// Fetch the names of all collections which the current user may read.
+    /// Each is mapped into a kafka topic.
+    pub async fn fetch_all_collection_names(&mut self) -> anyhow::Result<Vec<String>> {
+        let client = self.authenticated_client().await?.pg_client();
+        #[derive(serde::Deserialize)]
+        struct Row {
+            catalog_name: String,
+        }
+        let rows_builder = client
+            .from("live_specs_ext")
+            .eq("spec_type", "collection")
+            .select("catalog_name");
+
+        let items = flow_client::pagination::into_items::<Row>(rows_builder)
+            .map(|res| res.map(|Row { catalog_name }| catalog_name))
+            .try_collect()
+            .await
+            .context("listing current catalog specifications")?;
+
+        Ok(items)
     }
-    let rows_builder = client
-        .from("live_specs_ext")
-        .eq("spec_type", "collection")
-        .select("catalog_name");
-
-    let items = flow_client::pagination::into_items::<Row>(rows_builder)
-        .map(|res| res.map(|Row { catalog_name }| catalog_name))
-        .try_collect()
-        .await
-        .context("listing current catalog specifications")?;
+}
+
+impl TaskAuth {
+    pub async fn fetch_all_collection_names(&self) -> anyhow::Result<Vec<String>> {
+        Ok(self
+            .built_spec
+            .bindings
+            .iter()
+            .map(|b| {
+                serde_json::from_str::<crate::connector::DekafResourceConfig>(
+                    &b.resource_config_json,
+                )
+            })
+            .map_ok(|val| val.topic_name)
+            .collect::<Result<Vec<_>, _>>()?)
+    }
+
+    pub fn get_binding_for_topic(
+        &self,
+        topic_name: &str,
+    ) -> anyhow::Result<
+        Option<(
+            proto_flow::flow::materialization_spec::Binding,
+            DekafResourceConfig,
+        )>,
+    > {
+        Ok(self
```

Comment:
Either make this a hashmap or sort the vec and use binary search -- we've seen >3k bindings, this could potentially get slow